File: cli/python/prompd/commands/compile.py
"""Implementation of the `prompd compile` command."""
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Dict, Optional, Tuple

import click

from prompd.commands.common import console


@click.command(name="compile", context_settings=dict(ignore_unknown_options=True, allow_extra_args=True))
@click.argument("source", type=str)
@click.option(
    "--to",
    "output_format",
    default="markdown",
    help="Output format (markdown | provider-json [openai|anthropic] | provider-json:openai)",
)
@click.option("--to-markdown", is_flag=True, help="Shorthand for --to markdown")
@click.option(
    "--to-provider-json",
    type=click.Choice(["openai", "anthropic"]),
    help="Shorthand for --to provider-json <provider>",
)
@click.option("-p", "--param", multiple=True, help="Parameter in format key=value (repeat for multiple)")
@click.option(
    "-f",
    "--params-file",
    type=click.Path(exists=True, path_type=Path),
    multiple=True,
    help="Load parameters from JSON file (repeatable)",
)
@click.option("-o", "--output", type=click.Path(), help="Write compiled output to file")
@click.option("-v", "--verbose", is_flag=True, help="Verbose output")
@click.pass_context
def compile_command(
    ctx,
    source: str,
    output_format: str,
    to_markdown: bool,
    to_provider_json: Optional[str],
    param: Tuple[str, ...],
    params_file: Tuple[Path, ...],
    output: Optional[str],
    verbose: bool,
):
    """Compile a .prmd file or package reference to a target format."""
    try:
        source_path = Path(source)
        package_pattern = r"^(@[\w.-]+/[\w.-]+|[\w.-]+)@([\w.-]+)/(.+\.prmd)$"
        match = re.match(package_pattern, source)

        if match:
            package_ref = f"{match.group(1)}@{match.group(2)}"
            file_path_in_package = match.group(3)

            if verbose:
                console.print(f"[cyan]Resolving package:[/cyan] {package_ref}")
                console.print(f"[cyan]File path:[/cyan] {file_path_in_package}")

            from prompd.package_resolver import PackageResolver

            resolver = PackageResolver()
            try:
                package_path = resolver.resolve_package(package_ref)
                source_path = package_path / file_path_in_package

                if not source_path.exists():
                    console.print(f"[red]File not found in package:[/red] {file_path_in_package}")
                    console.print(f"[yellow]Package location:[/yellow] {package_path}")
                    raise SystemExit(1)

                if verbose:
                    console.print(f"[green]Resolved to:[/green] {source_path}")
            except Exception as exc:
                console.print(f"[red]Failed to resolve package:[/red] {exc}")
                raise SystemExit(1)
        elif not source_path.exists():
            if "@" in source and "/" not in source.split("@")[-1]:
                from prompd.package_resolver import PackageResolver

                resolver = PackageResolver()
                try:
                    package_path = resolver.resolve_package(source)
                    manifest_file = package_path / "manifest.json"
                    if manifest_file.exists():
                        with open(manifest_file, "r", encoding="utf-8") as mf:
                            manifest = json.load(mf)
                        main_file = manifest.get("main")
                        if main_file:
                            source_path = package_path / main_file
                            if verbose:
                                console.print(f"[green]Using main file:[/green] {main_file}")
                except Exception:
                    pass

            if not source_path.exists():
                console.print(f"[red]File not found:[/red] {source}")
                raise SystemExit(1)

        parameters: Dict[str, Any] = {}
        if params_file:
            for pf in params_file:
                try:
                    data = json.loads(Path(pf).read_text(encoding="utf-8"))
                    if isinstance(data, dict):
                        parameters.update(data)
                except Exception as exc:
                    console.print(f"[red]Error loading params file {pf}:[/red] {exc}")
                    raise SystemExit(1)

        if param:
            for kv in param:
                if "=" not in kv:
                    console.print(f"[red]Invalid parameter:[/red] {kv}. Use key=value")
                    raise SystemExit(1)
                key, value = kv.split("=", 1)

                try:
                    if value.strip().startswith(("{", "[")):
                        parameters[key] = json.loads(value)
                    elif value.lower() in ("true", "false"):
                        parameters[key] = value.lower() == "true"
                    elif value.replace(".", "").replace("-", "").isdigit() and value.count(".") <= 1:
                        parameters[key] = float(value) if "." in value else int(value)
                    else:
                        parameters[key] = value
                except json.JSONDecodeError:
                    parameters[key] = value

        if to_markdown:
            output_format = "markdown"
        elif to_provider_json:
            output_format = f"provider-json:{to_provider_json}"
        else:
            extra = list(getattr(ctx, "args", []) or [])
            if output_format.strip().lower() == "provider-json" and extra:
                next_tok = extra[0]
                if next_tok and not next_tok.startswith("-"):
                    output_format = f"provider-json:{next_tok}"
                    try:
                        ctx.args = extra[1:]
                    except Exception:
                        pass

        if verbose:
            try:
                console.print(
                    f"[dim]Compiling {source} -> {output_format} with params: {list(parameters.keys())}[/dim]"
                )
            except Exception:
                pass

        from prompd.compiler import PrompdCompiler

        compiler = PrompdCompiler()
        result = compiler.compile(
            source=source,
            output_format=output_format,
            parameters=parameters,
            output_file=Path(output) if output else None,
            verbose=verbose,
        )

        if output:
            try:
                console.print(f"[green]OK[/green] Compiled output written to {output}")
            except Exception:
                print(f"OK - Compiled output written to {output}")
        else:
            print(result)
    except SystemExit:
        raise
    except Exception as exc:
        try:
            console.print(f"[red]Error compiling:[/red] {exc}")
        except Exception:
            print(f"Error compiling: {exc}")
        raise SystemExit(1)


__all__ = ["compile_command"]

File: typescript/src/commands/compile.ts
/**
 * Compile Command
 *
 * Compiles a .prmd file or package reference using the 6-stage compilation pipeline.
 * Supports multiple output formats: markdown, OpenAI JSON, Anthropic JSON.
 */

import { Command } from 'commander';
import { readFileSync } from 'fs';
import { resolve } from 'path';
import { PrompdCompiler } from '../lib/compiler';

export function createCompileCommand(): Command {
  const cmd = new Command('compile');

  cmd
    .description('Compile a .prmd file or package reference with full template processing')
    .argument('<source>', '.prmd file path or package reference (@namespace/package@version)')
    .option('--to <format>', 'Output format: markdown | provider-json:openai | provider-json:anthropic', 'markdown')
    .option('--to-markdown', 'Shorthand for --to markdown')
    .option('--to-provider-json <provider>', 'Shorthand for --to provider-json:<provider>')
    .option('-p, --param <key=value>', 'Parameter in format key=value', collectParams, {})
    .option('-f, --params-file <file>', 'JSON parameter file')
    .option('-o, --output <file>', 'Output file path')
    .option('-v, --verbose', 'Verbose output with compilation details', false)
    .action(async (source: string, options: any) => {
      try {
        // Collect parameters from different sources
        let parameters: Record<string, any> = {};

        // Load from parameter file if provided
        if (options.paramsFile) {
          try {
            const fileContent = readFileSync(resolve(options.paramsFile), 'utf8');
            const fileParams = JSON.parse(fileContent);
            parameters = { ...parameters, ...fileParams };
          } catch (error) {
            console.error(`Error loading parameter file: ${error}`);
            process.exit(1);
          }
        }

        // Merge command line parameters (CLI takes precedence)
        parameters = { ...parameters, ...options.param };

        // Determine output format
        let outputFormat = options.to;

        if (options.toMarkdown) {
          outputFormat = 'markdown';
        } else if (options.toProviderJson) {
          outputFormat = `provider-json:${options.toProviderJson}`;
        }

        // Create compiler instance
        const compiler = new PrompdCompiler();

        if (options.verbose) {
          console.log(`Compiling: ${source}`);
          console.log(`Output format: ${outputFormat}`);
          console.log(`Parameters: ${Object.keys(parameters).length} parameter(s)`);
          console.log('');
        }

        // Compile with the 6-stage pipeline
        const result = await compiler.compile(source, {
          outputFormat,
          parameters,
          outputFile: options.output,
          verbose: options.verbose
        });

        // Output result (if not written to file)
        if (!options.output) {
          console.log('Compiled Content:');
          console.log('-'.repeat(70));
          console.log(result);
          console.log('-'.repeat(70));
        } else {
          console.log(`✓ Compiled content written to ${options.output}`);
        }
      } catch (error) {
        console.error(`Compilation error: ${error instanceof Error ? error.message : String(error)}`);
        process.exit(1);
      }
    });

  return cmd;
}

/**
 * Collect parameters from CLI arguments.
 */
function collectParams(value: string, previous: Record<string, any>): Record<string, any> {
  const [key, val] = value.split('=', 2);

  if (!val) {
    throw new Error(`Invalid parameter format: ${value}. Use key=value format.`);
  }

  // Try to parse as JSON for complex values
  let parsedValue: any = val;
  try {
    parsedValue = JSON.parse(val);
  } catch {
    // Not JSON, use as string
    parsedValue = val;
  }

  return { ...previous, [key]: parsedValue };
}
